plugins/camera/camera_impl.cpp

```diff
@@ -0,0 +1,739 @@
+        MAVLINK_MSG_ID_STORAGE_INFORMATION,
+        std::bind(&CameraImpl::process_storage_information, this, _1),
+        this);
+
+    _parent.register_mavlink_message_handler(
+        MAVLINK_MSG_ID_CAMERA_IMAGE_CAPTURED,
+        std::bind(&CameraImpl::process_camera_image_captured, this, _1),
+        this);
+
+    _parent.register_mavlink_message_handler(
+        MAVLINK_MSG_ID_CAMERA_SETTINGS,
+        std::bind(&CameraImpl::process_camera_settings, this, _1),
+        this);
+
+    _parent.register_mavlink_message_handler(
+        MAVLINK_MSG_ID_CAMERA_INFORMATION,
+        std::bind(&CameraImpl::process_camera_information, this, _1),
+        this);
+
+    _parent.send_command_with_ack_async(
+        MAV_CMD_REQUEST_CAMERA_INFORMATION,
+        MavlinkCommands::Params {1.0f, // request it
+                                 NAN, NAN, NAN, NAN, NAN, NAN},
+        nullptr,
+        MAV_COMP_ID_CAMERA);
+}
+
+void CameraImpl::deinit()
+{
+    delete _camera_definition;
+
+    _parent.unregister_all_mavlink_message_handlers(this);
+}
+
+void CameraImpl::enable() {}
+
+void CameraImpl::disable() {}
+
+Camera::Result CameraImpl::take_photo()
+{
+    std::lock_guard<std::mutex> lock(_capture.mutex);
+
+    return camera_result_from_command_result(
+               _parent.send_command_with_ack(
+                   MAV_CMD_IMAGE_START_CAPTURE,
+                   MavlinkCommands::Params {0.0f, // Reserved, set to 0
+                                            0.0f, // only picture, no interval
+                                            1.0f, // take only one picture
+                                            float(_capture.sequence++),
+                                            NAN, NAN, NAN},
+                   MAV_COMP_ID_CAMERA));
+}
+
+Camera::Result CameraImpl::start_photo_interval(float interval_s)
+{
+    if (!interval_valid(interval_s)) {
+        return Camera::Result::WRONG_ARGUMENT;
+    }
+
+    std::lock_guard<std::mutex> lock(_capture.mutex);
+
+    return camera_result_from_command_result(
+               _parent.send_command_with_ack(
+                   MAV_CMD_IMAGE_START_CAPTURE,
+                   MavlinkCommands::Params {0.0f, // Reserved, set to 0
+                                            interval_s,
+                                            0, // unlimited photos
+                                            float(_capture.sequence++),
+                                            NAN, NAN, NAN},
+                   MAV_COMP_ID_CAMERA));
+}
+
+Camera::Result CameraImpl::stop_photo_interval()
+{
+    return camera_result_from_command_result(
+               _parent.send_command_with_ack(
+                   MAV_CMD_IMAGE_STOP_CAPTURE,
+                   MavlinkCommands::Params {0.0f, // Reserved, set to 0
+                                            NAN, NAN, NAN, NAN, NAN, NAN},
+                   MAV_COMP_ID_CAMERA));
+}
+
+Camera::Result CameraImpl::start_video()
+{
```

Comment:
Don't we need to check whether video capture in progress ? Also, we need to make sure we're in Video mode. Same applies to async versions.

Reply:
I made a TODO.

Reply:
OK.